### PR TITLE
Add FDBKeeper implementation of IKeeper

### PR DESCRIPTION
**Pull request by zjial**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
As mentioned in [#48620](https://github.com/ClickHouse/ClickHouse/issues/48620), we developed FDBKeeper, an implementation of IKeeper that replaces Zookeeper. It does not change Clickhouse's original logic, but only implements the IKeeper interface on top of FoundationDB (by mapping Zookeeper data structure to FoundationDB KV pair and simulating session mechanism, etc.). FDBKeeper can be enabled by configuring `zookeeper.implementation`. A simple configuration example is as follows:

```
<clickhouse>
    <zookeeper>
        <implementation>fdbkeeper</implementation>
        <fdb_cluster>/etc/foundationdb/fdb.cluster</fdb_cluster>
    </zookeeper>
</clickhouse>
```

We conducted corresponding tests on Clickhouse using FDBKeeper. All functional tests passed. Resource consumption and performance tests showed that FDBKeeper can support clusters of the same scale with fewer resources than Zookeeper, and the performance is comparable to Zookeeper (the performance can be improved by increasing FDBKeeper resources).

Currently, the implementation is based on the April commit 44f05e3. We will rebase to the latest master branch if the community is willing to accept it.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by feixue1121** [comment]
@alexey-milovidov actually, We compared the resources used by FDB clusters and Keeper clusters. The results show that FDB cluster can use less resource than Keeper clusters and supports more ClickHouse instances. We'll post the experiment result later and hope the community will be interested in this feature.

**Comment by zjial** [comment]
Testing method

We conducted the tests by simulating user operations, including six concurrent `lwd + insert` operations and four concurrent `select` operation. The table engine used was `replicated database + replicated mergetree`.

Prerequisites

Each Clickhouse cluster has a dedicated 3-node Keeper cluster. Multiple Clickhouse clusters share an FDB cluster. Due to some reasons, the Clickhouse cluster here is 1 shard 1 replica.

Testing steps

We first tested the scenario of a Clickhouse cluster + 3-node Keeper cluster. Keeping other parameters unchanged, we used Fdbkeeper to replace Zookeeper and tested the scenario of Clickhouse clusters + Fdb cluster.

Testing results

We tested the scenario of 60 Clickhouse clusters + Fdb cluster and obtained an FDB cluster configuration that can make Clickhouse clusters run stably. Compared to Keeper, the FDB cluster requires fewer resources, roughly equivalent to 40 Keeper clusters.

In addition, we compared the rps of the IKeeper interface during the test, with one Clickhouse cluster + 3-node Keeper cluster as a reference. For example, the rps of the `list` interface for Keeper testing:

截屏2023-09-21 下午1 18 14

The rps of the 60 Clickhouse clusters + Fdb cluster testing is as follows:

截屏2023-09-14 下午5 35 28

The first part of the chart shows the rps of the interface when the number of Clickhouse clusters is small. The two drops in rps correspond to problems encountered during the test. Please refer to the results in the 17:00 and 18:15 time periods. The test results show that using Fdbkeeper does not degrade performance.

Conclusion

Based on the test results, Fdbkeeper does not degrade performance, and using FDBKeeper + FDB can support more Clickhouse instances than Keeper.

**Comment by alexey-milovidov** [comment]
Let's resolve conflicts and see the results of CI.

**Comment by zjial** [comment]
> Let's resolve conflicts and see the results of CI.

Sure, we will rebase to the latest master branch, which may take some time as we need to support the new IKeeper interface and resolve conflicts.

**Comment by qoega** [comment]
There is also a standalone keeper bench https://github.com/ClickHouse/ClickHouse/pull/48547

You can test Keeper/ZooKeeper/FDBKeeper with the different workload and compare results

**Comment by zjial** [comment]
Hi, @alexey-milovidov , we have rebased to the master branch, please take a look.

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 44dee2b86ca3ef06a140971b6468421fa6a70ce5 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54823/44dee2b86ca3ef06a140971b6468421fa6a70ce5/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by alexey-milovidov** [comment]
All tests have failed:

> Style Check — Check style failed. Check black failed. Check typos failed.

**Comment by zjial** [comment]
> All tests have failed:
> 
> > Style Check — Check style failed. Check black failed. Check typos failed.

Hi, @alexey-milovidov , we have formatted the style of some code and added support for some Keeper features. The workflow appears to require your approval.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , please approve the workflow.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , I noticed that every commit requires a maintainer's approval, and I am wondering if this is because of the `manual approve` label. Could you please tell me the reason for adding this label? I would also like to ask if this label can be removed.

**Comment by alexey-milovidov** [comment]
It's due to the changes in the `docker` directory. I approved.

File: contrib/fdb-c-cmake/CMakeLists.txt
Comment (alexey-milovidov) [inline contrib/fdb-c-cmake/CMakeLists.txt:57]: We don't allow cloning repositories at the build stage.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: contrib/fdb-c-cmake/CMakeLists.txt
Comment (zjial) [reply contrib/fdb-c-cmake/CMakeLists.txt:57]: Thank you for your review. Will it cause the compilation to fail? According to `run.log`, the repository seems to be successfully cloned. If not, we are going to continue running the workflow, and then solve the compilation problem (static linking and this problem) in parallel. In addition, since fdb depends on [mono](https://www.mono-project.com/) when compiling (which should be the reason of the failure of fast test compilation), we modified the docker file of fast test in the latest commit.

File: docker/test/fasttest/Dockerfile
Comment (alexey-milovidov) [inline docker/test/fasttest/Dockerfile:30]: This is a supply chain vulnerability (cannot be accepted), but let's see what will fail next.

**Comment by alexey-milovidov** [comment]
It does not build.

**Comment by zjial** [comment]
> It does not build.

The failure was due to it not being compiled with clang as expected. We fixed it and the subsequent `Wshadow` warning.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , let's continue the workflow please.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , let's continue the workflow please.

**Comment by alexey-milovidov** [comment]
Ok. Need to resolve conflicts.

**Comment by zjial** [comment]
> Ok. Need to resolve conflicts.

Ok.

**Comment by zjial** [comment]
Resolved. Additionally, it seems that Musl is only used when compiling clickhouse-keeper. Since FoundationDB does not support Musl and clickhouse-keeper does not need FDBKeeper, we disable it when Musl is used.

**Comment by alexey-milovidov** [comment]
Ok. Although we are planning to switch to Musl globally.

**Comment by zjial** [comment]
Thanks. Can you please tell me what is the reason, and is there an estimated implementation time?

**Comment by alexey-milovidov** [comment]
Currently, it does not build.

**Comment by zjial** [comment]
It seems that the `cache_dir` is `//.cache/ccache` when compiling binary packages, and then an error of `Permission denied` was reported when compiling fdb and creating the `//.cache/ccache/tmp` directory. Currently I temporarily disabled ccache, but I am a little curious why it has this problem.

**Comment by Sallery-X** [comment]
> ### Changelog category (leave one):
> * New Feature
> 
> ### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
> As mentioned in [#48620](https://github.com/ClickHouse/ClickHouse/issues/48620), we developed FDBKeeper, an implementation of IKeeper that replaces Zookeeper. It does not change Clickhouse's original logic, but only implements the IKeeper interface on top of FoundationDB (by mapping Zookeeper data structure to FoundationDB KV pair and simulating session mechanism, etc.). FDBKeeper can be enabled by configuring `zookeeper.implementation`. A simple configuration example is as follows:
> 
> ```
> <clickhouse>
>     <zookeeper>
>         <implementation>fdbkeeper</implementation>
>         <fdb_cluster>/etc/foundationdb/fdb.cluster</fdb_cluster>
>     </zookeeper>
> </clickhouse>
> ```
> 
> We conducted corresponding tests on Clickhouse using FDBKeeper. All functional tests passed. Resource consumption and performance tests showed that FDBKeeper can support clusters of the same scale with fewer resources than Zookeeper, and the performance is comparable to Zookeeper (the performance can be improved by increasing FDBKeeper resources).
> 
> Currently, the implementation is based on the April commit [44f05e3](https://github.com/ClickHouse/ClickHouse/commit/44f05e3fca7df6f093881cfd781fa6a676d98b53). We will rebase to the latest master branch if the community is willing to accept it.
> 
> ### Documentation entry for user-facing changes
> * [x]  Documentation is written (mandatory for new features)
> 
> > Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

Could I have your WeChat？ 
I would like to ask other questions

**Comment by zjial** [comment]
Hi, @Sallery-X , yeah, I sent you an email yesterday. Please check it.

**Comment by alexey-milovidov** [comment]
Now need to fix integration tests.

**Comment by zjial** [comment]
> Now need to fix integration tests.

Ok.

**Comment by zjial** [comment]
integration tests fixed.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , let's continue the workflow please.

**Comment by alexey-milovidov** [comment]
https://github.com/ClickHouse/ClickHouse/pull/56914

**Comment by alexey-milovidov** [comment]
Please merge with master.

**Comment by alexey-milovidov** [comment]
Let's enable FDB Keeper by default in this PR to check which tests will fail.

**Comment by zjial** [comment]
Thanks. Merge with master. Currently, FDBKeeper is enabled by default when compiling, and ZooKeeper is still used for testing (such as functional tests). Do you mean to modify all tests to replace ZooKeeper with FDBKeeper?

**Comment by alexey-milovidov** [comment]
Yes, so we will run all the tests with FDB.

**Comment by zjial** [comment]
Ok, it will take some time.

**Comment by zjial** [comment]
We have modified functional tests(**fast tests, stateless/stateful/stress tests**) and will continue to modify integration tests. Besides these tests, are there any other tests that need to be modified?

**Comment by zjial** [comment]
Modified some integration tests.

**Comment by zjial** [comment]
> It is pointless to duplicate the code of existing tests for FDB. Instead, the CI system can be switched to FDB in this PR to ensure all the existing tests have passed.

Thank you for your comment. Do you suggest that we modify the existing tests directly? For integration tests, we replaced `with_zookeeper` with `with_foundationdb` to enable `fdbkeeper`. To avoid modifying the existing tests, we duplicated it. If not, I would appreciate your suggestions for implementation.

**Comment by alexey-milovidov** [comment]
Do not modify or copy the existing tests, just switch to FoundationDB globally.

**Comment by zjial** [comment]
Currently, the functional tests switch to fdbkeeper by specifying `zookeeper.implementation` in `tests/config/install.sh`. To globally switch to fdbkeeper for integration tests, can we specify fdbkeeper in `tests/integration/helpers/cluster.py`? It is worth noting that replacing all of them may cause problems, as some of the tests do not seem to need fdbkeeper, such as `test_ttl_replicated` and `test_reload_zookeeper`.

**Comment by alexey-milovidov** [comment]
Our goal is to check - whether all the tests will pass if we replace Keeper with FoundationDB.

**Comment by zjial** [comment]
Hi, @alexey-milovidov , let's continue the workflow please.

**Comment by alexey-milovidov** [comment]
All tests have failed.

**Comment by alexey-milovidov** [comment]
The style check has failed.

**Comment by VanDarkholme7** [comment]
Hi! @alexey-milovidov   We can pass several timeout-related stateless tests locally. We had this problem before when running our CI checks until we switched to a separate and stable foundationdb cluster. Currently, fdb is started in the same container together with the test. If we want to start a separate container for the fdb cluster, what should we do?

**Comment by zjial** [comment]
> Hi! @alexey-milovidov We can pass several timeout-related stateless tests locally. We had this problem before when running our CI checks until we switched to a separate and stable foundationdb cluster. Currently, fdb is started in the same container together with the test. If we want to start a separate container for the fdb cluster, what should we do?

Hi, @alexey-milovidov , please let me know if you have any suggestions on this problem.

**Comment by alexey-milovidov** [comment]
Please resolve conflicts.

**Comment by alexey-milovidov** [comment]
Tests have failed.

**Comment by alexey-milovidov** [comment]
We can continue...

**Comment by xleoken** [comment]
👍

**Comment by alexey-milovidov** [comment]
No activity in three months.

**Comment by alexey-milovidov** [comment]
If someone wants this feature - open another pull request.

**Comment by Slach** [comment]
@zjial any news from your side about this feature?

**Comment by zjial** [comment]
> @zjial any news from your side about this feature?

I have recently left my previous company. If the company decides to continue developing and pushing this feature, my colleagues will take over and open another pull request, as @alexey-milovidov said. It seems that there may not be new commit in the short term.

**Comment by danthegoodman1** [comment]
That’s a massive shame

**Comment by alexey-milovidov** [comment]
No need to insult the contributors. Whether they want and have time to finish the implementation is up to them.
We don't demand contributions.

**Comment by danthegoodman1** [comment]
I wasn’t insulting anyone, i was just pointing out that I was very much looking forward to seeing this integrated, and that it’s sad to see that it didn’t get completed.

**Comment by danthegoodman1** [comment]
I wasn’t saying their acts were shameful, I’m saying the circumstance is unfortunate

**Comment by ylw510** [comment]
mark.

**Pull request closed without merging**
